File: apps/sketch2bim/scripts/create_razorpay_plans.py
"""
Script to create Razorpay Plans for subscription payments
Run this once to create Plans in your Razorpay account

Usage:
    # Option 1: Run from project root (if backend dependencies are installed globally)
    python scripts/create_razorpay_plans.py
    
    # Option 2: Run from backend directory (recommended)
    cd backend
    python ../scripts/create_razorpay_plans.py
    
    # Option 3: Run with virtual environment activated
    # source venv/bin/activate  # or `venv\Scripts\activate` on Windows
    python scripts/create_razorpay_plans.py
"""
import os
import sys
from pathlib import Path

# Add backend directory to path to import backend dependencies
backend_dir = Path(__file__).parent.parent / "backend"
if backend_dir.exists():
    sys.path.insert(0, str(backend_dir))

try:
    import razorpay
    from dotenv import load_dotenv
except ImportError as e:
    print(f"Error: Missing required package: {e}")
    print("\nPlease install backend dependencies:")
    print("  cd backend")
    print("  pip install -r requirements.txt")
    print("\nOr run this script from the backend directory:")
    print("  cd backend")
    print("  python ../scripts/create_razorpay_plans.py")
    sys.exit(1)

# Load app-specific production env
workspace_root = Path(__file__).resolve().parents[2]
app_env_path = workspace_root / "sketch2bim.env.production"
if app_env_path.exists():
    load_dotenv(app_env_path, override=False)
    print(f"Loaded environment from: {app_env_path}")
else:
    print(f"⚠️  Warning: sketch2bim.env.production not found at {app_env_path}")

RAZORPAY_KEY_ID = os.getenv('SKETCH2BIM_RAZORPAY_KEY_ID')
RAZORPAY_KEY_SECRET = os.getenv('SKETCH2BIM_RAZORPAY_KEY_SECRET')

if not RAZORPAY_KEY_ID or not RAZORPAY_KEY_SECRET:
    print("Error: SKETCH2BIM_RAZORPAY_KEY_ID and SKETCH2BIM_RAZORPAY_KEY_SECRET must be set")
    sys.exit(1)

# Initialize Razorpay client
client = razorpay.Client(auth=(RAZORPAY_KEY_ID, RAZORPAY_KEY_SECRET))

# Plan definitions
# Razorpay Plan structure: https://razorpay.com/docs/api/subscriptions/#create-a-plan
PLANS = [
    {
        "period": "weekly",
        "interval": 1,
        "item": {
            "name": "Week Pass Subscription",
            "description": "7-day access to Sketch-to-BIM",
            "amount": 129900,  # ₹1,299 in paise
            "currency": "INR"
        },
        "notes": {
            "tier": "week"
        }
    },
    {
        "period": "monthly",
        "interval": 1,
        "item": {
            "name": "Month Subscription",
            "description": "30-day access to Sketch-to-BIM",
            "amount": 349900,  # ₹3,499 in paise
            "currency": "INR"
        },
        "notes": {
            "tier": "month"
        }
    },
    {
        "period": "yearly",
        "interval": 1,
        "item": {
            "name": "Year Subscription",
            "description": "365-day access to Sketch-to-BIM",
            "amount": 2999900,  # ₹29,999 in paise
            "currency": "INR"
        },
        "notes": {
            "tier": "year"
        }
    }
]

def create_plan(plan_data):
    """Create a Razorpay Plan"""
    try:
        print(f"\nCreating plan: {plan_data['item']['name']}")
        print(f"  Period: {plan_data['period']}, Interval: {plan_data['interval']}")
        print(f"  Amount: ₹{plan_data['item']['amount'] / 100}")
        
        # Try creating the plan
        plan = client.plan.create(data=plan_data)
        
        print(f"✅ Created Plan: {plan['id']} - {plan['item']['name']}")
        return plan
    except razorpay.errors.BadRequestError as e:
        print(f"❌ Bad Request Error for {plan_data['item']['name']}:")
        print(f"   Status: {e.status_code if hasattr(e, 'status_code') else 'N/A'}")
        print(f"   Error: {str(e)}")
        if hasattr(e, 'error') and isinstance(e.error, dict):
            print(f"   Error details: {e.error}")
        return None
    except razorpay.errors.ServerError as e:
        print(f"❌ Server Error for {plan_data['item']['name']}: {str(e)}")
        return None
    except Exception as e:
        print(f"❌ Failed to create plan {plan_data['item']['name']}: {type(e).__name__}")
        print(f"   Error: {str(e)}")
        if hasattr(e, 'error'):
            print(f"   Error details: {e.error}")
        if hasattr(e, 'status_code'):
            print(f"   Status code: {e.status_code}")
        return None

def check_existing_plans():
    """Check if plans already exist"""
    try:
        plans = client.plan.all()
        existing = {}
        if plans and 'items' in plans:
            for plan in plans['items']:
                item_name = plan.get('item', {}).get('name', '')
                if 'Week Pass' in item_name:
                    existing['week'] = plan['id']
                elif 'Month Subscription' in item_name:
                    existing['month'] = plan['id']
                elif 'Year Subscription' in item_name:
                    existing['year'] = plan['id']
        return existing
    except Exception as e:
        print(f"⚠️  Could not check existing plans: {e}")
        return {}

def main():
    print("Creating Razorpay Plans for subscriptions...")
    print("=" * 60)
    print(f"Using Key ID: {RAZORPAY_KEY_ID[:10]}...")
    print(f"Mode: {'TEST' if 'test' in RAZORPAY_KEY_ID.lower() else 'LIVE'}")
    print("=" * 60)
    
    # Check for existing plans
    print("\nChecking for existing plans...")
    existing_plans = check_existing_plans()
    if existing_plans:
        print(f"Found {len(existing_plans)} existing plan(s):")
        for tier, plan_id in existing_plans.items():
            print(f"  {tier}: {plan_id}")
        print("\nSkipping creation of existing plans.")
    
    created_plans = existing_plans.copy()
    
    # Create missing plans
    for plan_data in PLANS:
        tier_key = plan_data['notes']['tier']
        if tier_key in existing_plans:
            print(f"\n⏭️  Skipping {tier_key} - plan already exists: {existing_plans[tier_key]}")
            continue
            
        plan = create_plan(plan_data)
        if plan:
            created_plans[tier_key] = plan['id']
    
    print("\n" + "=" * 60)
    print("Plan IDs (for environment variables):")
    print("=" * 60)
    
    if not created_plans:
        print("❌ No plans created or found. Please check the errors above.")
        print("\nTroubleshooting:")
        print("1. Verify your API keys are correct")
        print("2. Check if your Razorpay account has subscription features enabled")
        print("3. Try creating plans manually in Razorpay Dashboard")
        return
    
    for tier in ['week', 'month', 'year']:
        if tier in created_plans:
            print(f"RAZORPAY_PLAN_{tier.upper()}={created_plans[tier]}")
        else:
            print(f"RAZORPAY_PLAN_{tier.upper()}=  # MISSING - create manually")
    
    print("\n" + "=" * 60)
    print("Copy these to your Render environment variables:")
    print("=" * 60)
    for tier in ['week', 'month', 'year']:
        if tier in created_plans:
            print(f"RAZORPAY_PLAN_{tier.upper()}={created_plans[tier]}")
    
    if len(created_plans) == 3:
        print("\n✅ All plans created successfully!")
    else:
        print(f"\n⚠️  Only {len(created_plans)}/3 plans created. Please create the missing ones manually.")

if __name__ == "__main__":
    main()

File: apps/reframe/scripts/create_razorpay_plans.ts
#!/usr/bin/env tsx

/**
 * Script to create Razorpay Plans for Reframe subscription payments
 * Run this once to create Plans in your Razorpay account
 * 
 * Usage:
 *   npx tsx scripts/create_razorpay_plans.ts
 * 
 * Or add to package.json:
 *   "scripts": {
 *     "create-razorpay-plans": "tsx scripts/create_razorpay_plans.ts"
 *   }
 */

import Razorpay from "razorpay";
import dotenv from "dotenv";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load app-specific .env.production
const appEnvPath = path.resolve(__dirname, "../../reframe.env.production");
if (fs.existsSync(appEnvPath)) {
  dotenv.config({ path: appEnvPath, override: false });
  console.log("Loaded environment from: reframe.env.production");
}

const RAZORPAY_KEY_ID = process.env.REFRAME_RAZORPAY_KEY_ID;
const RAZORPAY_KEY_SECRET = process.env.REFRAME_RAZORPAY_KEY_SECRET;

if (!RAZORPAY_KEY_ID || !RAZORPAY_KEY_SECRET) {
  console.error("Error: REFRAME_RAZORPAY_KEY_ID and REFRAME_RAZORPAY_KEY_SECRET must be set");
  console.error("They should be in: reframe.env.production");
  process.exit(1);
}

// Initialize Razorpay client
const razorpay = new Razorpay({
  key_id: RAZORPAY_KEY_ID,
  key_secret: RAZORPAY_KEY_SECRET,
});

// Plan definitions for Reframe
// Unified pricing shared across all apps (ASK, Reframe, Sketch2BIM)
// Week: ₹1,299 = 129900 paise
// Monthly: ₹3,499 = 349900 paise
// Yearly: ₹29,999 = 2999900 paise
const PLANS = [
  {
    period: "weekly" as const,
    interval: 1,
    item: {
      name: "Week Pro - Reframe",
      description: "7-day unlimited access to Reframe AI",
      amount: 129900, // ₹1,299 in paise
      currency: "INR",
    },
    notes: {
      tier: "week",
      app: "reframe",
    },
  },
  {
    period: "monthly" as const,
    interval: 1,
    item: {
      name: "Monthly Pro - Reframe",
      description: "30-day unlimited access to Reframe AI",
      amount: 349900, // ₹3,499 in paise
      currency: "INR",
    },
    notes: {
      tier: "monthly",
      app: "reframe",
    },
  },
  {
    period: "yearly" as const,
    interval: 1,
    item: {
      name: "Yearly Pro - Reframe",
      description: "365-day unlimited access to Reframe AI",
      amount: 2999900, // ₹29,999 in paise
      currency: "INR",
    },
    notes: {
      tier: "yearly",
      app: "reframe",
    },
  },
];

interface PlanData {
  period: "weekly" | "monthly" | "yearly";
  interval: number;
  item: {
    name: string;
    description: string;
    amount: number;
    currency: string;
  };
  notes: {
    tier: string;
    app: string;
  };
}

async function createPlan(planData: PlanData): Promise<any | null> {
  try {
    console.log(`\nCreating plan: ${planData.item.name}`);
    console.log(`  Period: ${planData.period}, Interval: ${planData.interval}`);
    console.log(`  Amount: ₹${planData.item.amount / 100}`);

    const plan = await razorpay.plans.create({
      period: planData.period,
      interval: planData.interval,
      item: planData.item,
      notes: planData.notes,
    });

    console.log(`✅ Created Plan: ${plan.id} - ${plan.item.name}`);
    return plan;
  } catch (error: any) {
    console.error(`❌ Failed to create plan ${planData.item.name}:`);
    if (error.statusCode) {
      console.error(`   Status: ${error.statusCode}`);
    }
    if (error.error) {
      console.error(`   Error: ${JSON.stringify(error.error, null, 2)}`);
    } else {
      console.error(`   Error: ${error.message || String(error)}`);
    }
    return null;
  }
}

async function checkExistingPlans(): Promise<Record<string, string>> {
  try {
    const plans = await razorpay.plans.all();
    const existing: Record<string, string> = {};

    if (plans.items) {
      for (const plan of plans.items) {
        const itemName = plan.item?.name || "";
        if (itemName.includes("Week Pro - Reframe")) {
          existing.week = plan.id;
        } else if (itemName.includes("Monthly Pro - Reframe")) {
          existing.monthly = plan.id;
        } else if (itemName.includes("Yearly Pro - Reframe")) {
          existing.yearly = plan.id;
        }
      }
    }

    return existing;
  } catch (error: any) {
    console.error(`⚠️  Could not check existing plans: ${error.message || String(error)}`);
    return {};
  }
}

async function main() {
  console.log("Creating Razorpay Plans for Reframe subscriptions...");
  console.log("=".repeat(60));
  console.log(`Using Key ID: ${RAZORPAY_KEY_ID.substring(0, 10)}...`);
  console.log(`Mode: ${RAZORPAY_KEY_ID.includes("test") ? "TEST" : "LIVE"}`);
  console.log("=".repeat(60));

  // Check for existing plans
  console.log("\nChecking for existing plans...");
  const existingPlans = await checkExistingPlans();
  if (Object.keys(existingPlans).length > 0) {
    console.log(`Found ${Object.keys(existingPlans).length} existing plan(s):`);
    for (const [tier, planId] of Object.entries(existingPlans)) {
      console.log(`  ${tier}: ${planId}`);
    }
    console.log("\nSkipping creation of existing plans.");
  }

  const createdPlans: Record<string, string> = { ...existingPlans };

  // Create missing plans
  for (const planData of PLANS) {
    const tierKey = planData.notes.tier;
    if (existingPlans[tierKey]) {
      console.log(`\n⏭️  Skipping ${tierKey} - plan already exists: ${existingPlans[tierKey]}`);
      continue;
    }

    const plan = await createPlan(planData);
    if (plan) {
      createdPlans[tierKey] = plan.id;
    }
  }

  console.log("\n" + "=".repeat(60));
  console.log("Plan IDs (for environment variables):");
  console.log("=".repeat(60));

  if (Object.keys(createdPlans).length === 0) {
    console.error("❌ No plans created or found. Please check the errors above.");
    console.log("\nTroubleshooting:");
    console.log("1. Verify your API keys are correct in reframe.env.production");
    console.log("2. Check if your Razorpay account has subscription features enabled");
    console.log("3. Try creating plans manually in Razorpay Dashboard");
    process.exit(1);
  }

  for (const tier of ["week", "monthly", "yearly"]) {
    if (createdPlans[tier]) {
      const envVarName = tier === "week" ? "REFRAME_RAZORPAY_PLAN_WEEK" :
                        tier === "monthly" ? "REFRAME_RAZORPAY_PLAN_MONTH" : 
                        "REFRAME_RAZORPAY_PLAN_YEAR";
      console.log(`${envVarName}=${createdPlans[tier]}`);
    } else {
      const envVarName = tier === "week" ? "REFRAME_RAZORPAY_PLAN_WEEK" :
                        tier === "monthly" ? "REFRAME_RAZORPAY_PLAN_MONTH" : 
                        "REFRAME_RAZORPAY_PLAN_YEAR";
      console.log(`${envVarName}=  # MISSING - create manually`);
    }
  }

  console.log("\n" + "=".repeat(60));
  console.log("Copy these to your reframe.env.production:");
  console.log("=".repeat(60));
  for (const tier of ["week", "monthly", "yearly"]) {
    if (createdPlans[tier]) {
      const envVarName = tier === "week" ? "REFRAME_RAZORPAY_PLAN_WEEK" :
                        tier === "monthly" ? "REFRAME_RAZORPAY_PLAN_MONTH" : 
                        "REFRAME_RAZORPAY_PLAN_YEAR";
      console.log(`${envVarName}=${createdPlans[tier]}`);
    }
  }

  if (Object.keys(createdPlans).length === 3) {
    console.log("\n✅ All plans created successfully!");
  } else {
    console.log(`\n⚠️  Only ${Object.keys(createdPlans).length}/3 plans created. Please create the missing ones manually.`);
  }
}

main().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});

